### PR TITLE
Refactor SQL execution to use prepared statements

### PR DESCRIPTION
**Pull request by Peterburnett**
SQL could be broken with values from a variable like `'`. Each variable is now processed as a parameter, and the whole SQL processed as a prepared statement

### CODE REVIEW

File: classes/local/step/flow_sql.php [hunk 1]
```diff
@@ -73,13 +73,13 @@ public function execute($input = null) {
 
         // Construct the query.
         $variables = $this->get_variables();
-        $config = $variables->get('config');
-        $sql = $this->evaluate_expressions($config->sql);
+        $config = $variables->get_raw('config');
+        list($sql, $params) = $this->evaluate_expressions($config->sql);
```
Comment (keevan) [inline classes/local/step/flow_sql.php:77]:
```suggestion
        [$sql, $params] = $this->evaluate_expressions($config->sql);
```
An alternative which is shorter, and used more in modern languages (e.g. JS)

File: classes/local/step/reader_sql.php [hunk 2]
```diff
@@ -76,19 +76,20 @@ public function is_concurrency_supported() {
      * @throws \moodle_exception
      */
     public function get_iterator(): iterator {
-        $query = $this->construct_query();
+        list($query, $params) = $this->construct_query();
```
Comment (keevan) [inline classes/local/step/reader_sql.php:79]: Same as above
```suggestion
        [$query, $params] = $this->construct_query();
```

File: classes/local/step/flow_sql.php (hunk 1 above)
Comment (Peterburnett) [reply classes/local/step/flow_sql.php:77]: Ah yeah I forgot the deconstruction got cleaner a while back. neat

File: classes/local/step/reader_sql.php [hunk 3]
```diff
@@ -183,9 +184,9 @@ function ($message, $e = null) use ($rawsql, &$errormsg) {
         }
 
         // Evalulate any remaining expressions as per normal.
-        $finalsql = $this->evaluate_expressions($finalsql);
+        list($finalsql, $params) = $this->evaluate_expressions($finalsql);
 
-        return $finalsql;
+        return [$finalsql, $params];
```
Comment (keevan) [inline classes/local/step/reader_sql.php:189]: This doesn't look necessary to me. I might be missing something, are you able to clarify?

Comment (Peterburnett) [reply classes/local/step/reader_sql.php]: its very not necessary. Didnt spot it at the time, as I was just following blind refactoring the touch points at the time

**Review by keevan — APPROVED** [review]
Thanks Peter. Looks good.

File: classes/local/step/flow_sql.php [hunk 4]
```diff
@@ -73,13 +73,13 @@ public function execute($input = null) {
 
         // Construct the query.
         $variables = $this->get_variables();
-        $config = $variables->get('config');
-        $sql = $this->evaluate_expressions($config->sql);
+        $config = $variables->get_raw('config');
+        [$sql, $params] = $this->evaluate_expressions($config->sql);
```
Comment (brendanheywood) [inline classes/local/step/flow_sql.php:77]: Should this pr touch lang packs / docs?

Comment (keevan) [reply classes/local/step/flow_sql.php:77]: Potentially here: 
https://github.com/catalyst/moodle-tool_dataflows/blob/MOODLE_35_STABLE/VARIABLES.md

Or in the SQL steps docs

**Pull request merged**